// jsii/test/enums.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const lib_1 = require("../lib");
// ----------------------------------------------------------------------
test('test parsing enum with two members and no values', async () => {
    const assembly = await lib_1.sourceToAssemblyHelper(`
    export enum Foo {
      BAR,
      BAZ
    }
  `);
    expect(assembly.types['testpkg.Foo']).toEqual({
        assembly: 'testpkg',
        fqn: 'testpkg.Foo',
        kind: 'enum',
        members: [{ name: 'BAR' }, { name: 'BAZ' }],
        locationInModule: { filename: 'index.ts', line: 2 },
        name: 'Foo',
        symbolId: 'index:Foo',
    });
});
// ----------------------------------------------------------------------
test('test parsing enum with two members and assigned values', async () => {
    const assembly = await lib_1.sourceToAssemblyHelper(`
    export enum Foo {
      BAR = 'Bar',
      BAZ = 'Baz'
    }
  `);
    expect(assembly.types['testpkg.Foo']).toEqual({
        assembly: 'testpkg',
        fqn: 'testpkg.Foo',
        kind: 'enum',
        members: [{ name: 'BAR' }, { name: 'BAZ' }],
        locationInModule: { filename: 'index.ts', line: 2 },
        name: 'Foo',
        symbolId: 'index:Foo',
    });
});
