// jsii-pacmak/lib/targets/go/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tarballName = exports.substituteReservedWords = exports.getParamDependencies = exports.getMemberDependencies = exports.flatMap = exports.goPackageNameForAssembly = exports.findTypeInTree = void 0;
/*
 * Recursively search module for type with fqn
 */
function findTypeInTree(module, fqn) {
    const result = module.types.find((t) => t.type.fqn === fqn);
    if (result) {
        return result;
    }
    return module.submodules.reduce((accum, sm) => {
        return accum !== null && accum !== void 0 ? accum : findTypeInTree(sm, fqn);
    }, undefined);
}
exports.findTypeInTree = findTypeInTree;
/*
 * Format NPM package names as idiomatic Go module name
 */
function goPackageNameForAssembly(assembly) {
    var _a, _b;
    const config = (_b = (_a = assembly.targets) === null || _a === void 0 ? void 0 : _a.go) !== null && _b !== void 0 ? _b : {};
    if (config.packageName) {
        return config.packageName;
    }
    return assembly.name.replace(/[^a-z0-9.]/gi, '').toLowerCase();
}
exports.goPackageNameForAssembly = goPackageNameForAssembly;
function flatMap(collection, mapper) {
    return collection
        .map(mapper)
        .reduce((acc, elt) => acc.concat(elt), new Array());
}
exports.flatMap = flatMap;
/*
 * Return module dependencies of a class or interface members
 */
function getMemberDependencies(members) {
    var _a, _b;
    const deps = new Array();
    for (const member of members) {
        deps.push(...((_b = (_a = member.reference) === null || _a === void 0 ? void 0 : _a.dependencies) !== null && _b !== void 0 ? _b : []));
    }
    return deps;
}
exports.getMemberDependencies = getMemberDependencies;
function getParamDependencies(methods) {
    var _a, _b;
    const dependencies = [];
    for (const method of methods) {
        for (const param of method.parameters) {
            dependencies.push(...((_b = (_a = param.reference) === null || _a === void 0 ? void 0 : _a.dependencies) !== null && _b !== void 0 ? _b : []));
        }
    }
    return dependencies;
}
exports.getParamDependencies = getParamDependencies;
const RESERVED_WORDS = {
    break: 'break_',
    default: 'default_',
    func: 'func_',
    interface: 'interface_',
    select: 'select_',
    case: 'case_',
    defer: 'defer_',
    go: 'go_',
    map: 'map_',
    struct: 'struct_',
    chan: 'chan_',
    else: 'else_',
    goto: 'goto_',
    package: 'package_',
    switch: 'switch_',
    const: 'const_',
    fallthrough: 'fallthrough_',
    if: 'if_',
    range: 'range_',
    type: 'type_',
    continue: 'continue_',
    for: 'for_',
    import: 'import_',
    return: 'return_',
    var: 'var_',
    _: '_arg',
};
/*
 * Sanitize reserved words
 */
function substituteReservedWords(name) {
    return RESERVED_WORDS[name] || name;
}
exports.substituteReservedWords = substituteReservedWords;
/**
 * Computes a safe tarball name for the provided assembly.
 *
 * @param assm the assembly.
 *
 * @returns a tarball name.
 */
function tarballName(assm) {
    const name = assm.name.replace(/^@/, '').replace(/\//g, '-');
    return `${name}-${assm.version}.tgz`;
}
exports.tarballName = tarballName;
