// jsii-pacmak/lib/targets/go/runtime/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JSII_FQN = exports.JSII_ANY = exports.JSII_OVERRIDE = exports.JSII_SSET_FUNC = exports.JSII_SET_FUNC = exports.JSII_SGET_FUNC = exports.JSII_GET_FUNC = exports.JSII_SINVOKE_VOID_FUNC = exports.JSII_SINVOKE_FUNC = exports.JSII_INVOKE_VOID_FUNC = exports.JSII_INVOKE_FUNC = exports.JSII_CREATE_FUNC = exports.JSII_INIT_ALIAS = exports.JSII_INIT_FUNC = exports.JSII_INIT_PACKAGE = exports.JSII_RT_ALIAS = exports.JSII_RT_PACKAGE_NAME = exports.JSII_RT_MODULE_NAME = void 0;
// JSII go runtime module name
exports.JSII_RT_MODULE_NAME = 'github.com/aws/jsii-runtime-go';
exports.JSII_RT_PACKAGE_NAME = `${exports.JSII_RT_MODULE_NAME}/runtime`;
exports.JSII_RT_ALIAS = '_jsii_';
// Jsii initializer package name
exports.JSII_INIT_PACKAGE = 'jsii';
// Function to initialize a jsii-generated module
exports.JSII_INIT_FUNC = 'Initialize';
// Alias used for the jsii init
exports.JSII_INIT_ALIAS = '_init_';
// Function to make create request
exports.JSII_CREATE_FUNC = `${exports.JSII_RT_ALIAS}.Create`;
// JSII invoke request
exports.JSII_INVOKE_FUNC = `${exports.JSII_RT_ALIAS}.Invoke`;
exports.JSII_INVOKE_VOID_FUNC = `${exports.JSII_RT_ALIAS}.InvokeVoid`;
// JSII static invoke
exports.JSII_SINVOKE_FUNC = `${exports.JSII_RT_ALIAS}.StaticInvoke`;
exports.JSII_SINVOKE_VOID_FUNC = `${exports.JSII_RT_ALIAS}.StaticInvokeVoid`;
// JSII get request
exports.JSII_GET_FUNC = `${exports.JSII_RT_ALIAS}.Get`;
// JSII static get request
exports.JSII_SGET_FUNC = `${exports.JSII_RT_ALIAS}.StaticGet`;
// JSII set request
exports.JSII_SET_FUNC = `${exports.JSII_RT_ALIAS}.Set`;
// JSII static set request
exports.JSII_SSET_FUNC = `${exports.JSII_RT_ALIAS}.StaticSet`;
// MISC types & functions
// Jsii override struct type
exports.JSII_OVERRIDE = `${exports.JSII_RT_ALIAS}.Override`;
// Jsii Any mock
exports.JSII_ANY = `${exports.JSII_RT_ALIAS}.Any`;
// Jsii FQN type
exports.JSII_FQN = `${exports.JSII_RT_ALIAS}.FQN`;
